design/re-introspection-interface.md

```diff
@@ -0,0 +1,76 @@
+# re-introspection API
+
+## Status
+
+implementable
+
+## Summary
+
+A declarative API is proposed to request the baremetal operator to
+re-introspect a `Ready` BareMetalHost.
+
+## Motivation
+
+We would like to have an interface to allow a user to re-gather hardware
+inventory of a `Ready` BareMetalHost when a hardware replacement is made.
+When a user of the underlying infrastructure makes some changes to the actual
+server (e.g. replace or add NIC, disk, etc.), the latest hardware inventory
+including those changes need to be re-collected and updated on the spec of the
+corresponding BareMetalHost object without having to delete it.
+
+Implementation of this proposal is based on using annotation (similar to
+[Reboot API](https://github.com/metal3-io/metal3-docs/blob/master/design/reboot-interface.md))
+to request re-introspection of a `Ready` BareMetalHost.
+Once the annotation is set on BareMetalHost, the baremetal operator will
+request hardware introspection of the host from
+Ironic.
+
+### Goals
+
+- A declarative API to perform re-introspection
+- Use this API for future Metal³ remediation controller
+
+### Non-Goals
+
+- Automated mechanism to trigger re-introspection
+
+## Proposal
+
+We follow Reboot API implementation and expect to implement the similar
+annotation based interface for a user to trigger re-introspection.
+
+`reintrospection.metal3.io` annotation form on BareMetalHost object
+triggers the controller to query Ironic for introspection of a host. This form
+has set-and-forget semantics and controller removes the annotation once
+introspection is completed.
+While the host is being re-introspected, the BareMetalHost will stay in
+`Inspecting` state until the process is completed.
+
+Re-introspection API can be requested only when BareMetalHost is in `Ready`
+state. If a re-introspection request is made while BareMetalHost is any other
+state than `Ready`, the request will be ignored, i.e., annotation will
+not be set on BareMetalHost object. This is important in order to not reboot a
```

Comment:
If the user sets the annotation and we're going to ignore it, is that an error? Do we remove the annotation, or leave it there? What happens to a provisioned host with the annotation that is then deprovisioned? Would that trigger introspection again?

Reply:
Could it be that we leave the annotation there until the host is in a state where we can perform the operation ? How would you otherwise prevent the annotation to be set ?

Reply:
I think we should not introspect a provisioned BMH, so that we don’t end up with deprovisioning it.

I can think of three different options about what to do with the annotation set on a provisioned BMH.

1. If the user sets the annotation on a provisioned BMH nothing will happen from the user’s perspective, while from the operator's side we still remove the annotation without calling introspection. 
2. Keep the annotation until `actionDeprovisioning` is called. After deprovisioning is completed, call the introspection and remove the annotation.
3. Keep the annotation until ` actionDeprovisioning` is called. After deprovisioning is completed, remove the annotation but don’t call the introspection.  

Third option is somehow similar to Reboot API, in a sense that you can’t ask for a power off while the BMH is in Ready state, but if you still set the annotation (suffixed) on a Ready BMH then nothing will happen but the annotation will be kept and removed during the provisioning If I’m not wrong.

However, the second option makes more sense to me and seems safe because introspection will be called only after deprovisioning and the user will not have to delete the BMH in order to keep the hardware details up to date.

Reply:
@dhellmann Does second option make sense to you?

Reply:
Another option could be an admission webhook which validates that the annotation being set only on `Ready` state.
Otherwise, a clear error is being returned to the user.

Reply:
I would suggest to ignore the annotation unless we are in the Ready or Inspecting states.
In the Ready state, if the annotation exists (and ExternallyProvisioned is not set) then move to the Inspecting state.
In the Inspecting state, delete the annotation if it exists.
(This is another way of describing option 2.)

The reason option 3 makes sense for the reboot one is that when you provision the server, it *has* been rebooted so that counts as success (and we go ahead and remove the annotation). The same does not apply to inspection, so that wouldn't be appropriate here.

Reply:
The admission web hook @n1r1 mentions would certainly produce the clearest result. Although we (or at least I) have been trying to avoid adding web hooks.

Maybe this is really something else to solve as part of the API decomposition? If the inspection data is its own resource, deleting it could trigger reprovisioning.

Reply:
> Maybe this is really something else to solve as part of the API decomposition? If the inspection data is its own resource, deleting it could trigger reprovisioning.

Yeah, I think that could be the way to go long-term (which is why I asked about just deleting the hardware details from the CR). But given that the data is in the Status, I think the annotation is a reasonable way of approximating the same semantics in the short term.

Reply:
I agree with @n1r1 's point regarding webhooks, but as @zaneb mentioned that could be something as a future enhancement (I can include it in the proposal). Maybe for now we can live with annotations, but soon we will anyway need to somehow formalize those annotations we use for introspection & reboot & power off/on.

Reply:
> I would suggest to ignore the annotation unless we are in the Ready or Inspecting states.
> In the Ready state, if the annotation exists (and ExternallyProvisioned is not set) then move to the Inspecting state.
> In the Inspecting state, delete the annotation if it exists.
> (This is another way of describing option 2.)

Thanks, will add it.

> The reason option 3 makes sense for the reboot one is that when you provision the server, it _has_ been rebooted so that counts as success (and we go ahead and remove the annotation). The same does not apply to inspection, so that wouldn't be appropriate here.

Yeah, make sense.

Reply:
Could we also consider the case where a host previously failed to inspect, and we want to retry?  Currently there's no way to recover from that situation (temporary network outage etc) without re-creating the BMH, so using this annotation could be more convenient in that case?

Reply:
@hardys tbh I think we should probably just change that behaviour so it always tries again.